pkg/kncloudevents/message_sender.go

```diff
@@ -189,6 +189,6 @@ func RetryConfigFromDeliverySpec(spec duckv1.DeliverySpec) (RetryConfig, error)
 	return retryConfig, nil
 }
 
-func checkRetry(_ context.Context, resp *nethttp.Response, _ error) (bool, error) {
-	return resp != nil && resp.StatusCode >= 300, nil
+func checkRetry(_ context.Context, resp *nethttp.Response, err error) (bool, error) {
+	return !(resp != nil && resp.StatusCode < 300), err
```

Comment:
IMO we should not retry on any error... Some of them are reasonable to not retry, like `dial tcp timeout` when the tcp address it's not reachable, or dns errors... My feeling is that it should be configurable which errors to retry and which not, because otherwise you bloat the dispatchers always in retrying, when there is no need (because maybe the pointed service doesn't exist at all)....

What are people feelings about that? @matzew, @vaikas, @lionelvillard?

Reply:
I think retrying on any error is fine. I'd rather not have the user deal with having to define too many of the errors, network, dns, etc. If they want retry, seems reasonable to expect any error to be retried and not just errors that happen at the protocol level.

Reply:
On k8s, pods may be started and deleted for many reasons (scale-up/down, node drain, node restart, ...), and not all addressables will implement proper readiness and graceful shutdown, so it may be quite common to get all kinds of errors during normal operation, and IMHO user will expect that setting "retries" will make the dispatcher retry in these kinds of situations...

Another thing to consider is that doing retries on network errors may cause a duplicate event to be delivered (which seems allowd by the cloudevents spec, https://github.com/cloudevents/spec/blob/v1.0/spec.md#id ) , so perhaps some users may prefer for events to be lost instead of potentially delivering twice... (of course, they still can set retries to 0 then...)

Reply:
> Another thing to consider is that doing retries on network errors may cause a duplicate event to be delivered (which seems allowd by the cloudevents spec, https://github.com/cloudevents/spec/blob/v1.0/spec.md#id ) , so perhaps some users may prefer for events to be lost instead of potentially delivering twice... (of course, they still can set retries to 0 then...)

Well, but this ends up in the discussions of "exactly once", which atm is out of scope of eventing

Reply:
I think retrying is a reasonable default, then if there are interests for specific configurations we can always add them.

Reply:
I agree with @slinkydeveloper that not every error should be retried.  fwiw... this is the logic in the "distributed" KafkaChannel dispatcher for determining when to retry.   Also, we only retry some of the following because of the broker munging the actual responses(400), and the retry test expecting them(429)...

```
	//
	// Note - Normally we would NOT want to retry 400 responses, BUT the knative-eventing
	//        filter handler (due to CloudEvents SDK V1 usage) is swallowing the actual
	//        status codes from the subscriber and returning 400s instead.  Once this has,
	//        been resolved we can remove 400 from the list of codes to retry.
	//
	if statusCode >= 500 || statusCode == 400 || statusCode == 404 || statusCode == 429 {
		logger.Warn("Failed To Send Message To Subscriber Service - Retrying")
		return true, nil
	} else if statusCode >= 300 && statusCode <= 399 {
		logger.Warn("Failed To Send Message To Subscriber Service - Not Retrying")
		return false, nil
	} else if statusCode == -1 {
		logger.Warn("No StatusCode Detected In Error - Retrying")
		return true, nil
	}

	// Do Not Retry 1XX, 2XX, & Most 4XX StatusCode Responses
	return false, nil
```

Reply:
We're discussing network errors, not status codes, for status codes, there is a separate issue: https://github.com/knative/eventing/issues/2411

Reply:
+1 for retry in error. Network error may happen for various reason IMO. If we have concerns about too many retries, we can always use other backoff strategy, eg exponential.

Reply:
+1 for retry in error.

Reply:
> I think retrying on any error is fine. I'd rather not have the user deal with having to define too many of the errors, network, dns, etc. If they want retry, seems reasonable to expect any error to be retried and not just errors that happen at the protocol level.

that would be my thinking here too